[GPII-3425]: Move all TLS primitives to helm-tls module

natarajaya opened the pull request:
This is required to handle situations when we only need to fetch generated TLS assets from TF state without any other interactions with existing resources in the cluster.
PR for infra part is pending.

stepanstipl commented on modules/helm-tls/main.tf:
```diff
@@ -0,0 +1,150 @@
+resource "local_file" "ca_key" {
+  filename = "${var.secrets_dir}/${local.tls_dir}/${var.helm_dir_name}/ca.key.pem"
+  content  = "${tls_private_key.root.private_key_pem}"
+}
```
Why is this block here, if it's commented out?

stepanstipl commented on modules/helm-tls/variables.tf:
```diff
@@ -0,0 +1,18 @@
+# This module will generate TLS assets that may later be used
+# To secure Tiller installation by helm-initializer module with TLS authentication.
+# By default, TLS assets will be generated to a following directory:
+# ${TF_VAR_secrets_dir}/${custom_tls_dir:-tiller-namespace}/${helm_dir_name}
```
`${TF_VAR_secrets_dir}` shouldn't be just `{secrets_dir}`?

stepanstipl commented on modules/helm-initializer/main.tf:
```diff
@@ -61,9 +56,9 @@ resource "null_resource" "install_tiller" {
-        --tiller-tls-key=${local_file.tiller_key.filename} \
+        --tls-ca-cert=${local.tls_dir}/ca.cert.pem \
+        --tiller-tls-cert=${local.tls_dir}/tiller.cert.pem \
+        --tiller-tls-key=${local.tls_dir}/tiller.key.pem \
```
What about keeping the actual filenames as variables? -> They would be potentially customisable + we would avoid repeating hardcoded names at 3 diff places.

natarajaya replied:
Sounds good!

stepanstipl commented on modules/helm-tls/main.tf:
```diff
@@ -0,0 +1,150 @@
+# ------------------------------------------------------------------------------
+
+locals {
+  tls_dir       = "${var.custom_tls_dir == "" ? var.tiller_namespace : var.custom_tls_dir}"
```
This is not consistent with the other module: `tls_dir       = "${var.secrets_dir}/${var.custom_tls_dir == "" ? var.tiller_namespace : var.custom_tls_dir}/${var.helm_dir_name}"`

natarajaya replied:
Nice catch! Updated.

stepanstipl commented on modules/helm-tls/main.tf:
```diff
@@ -0,0 +1,150 @@
+
+resource "local_file" "ca_cert" {
+  filename = "${var.secrets_dir}/${local.tls_dir}/${var.helm_dir_name}/ca.cert.pem"
+  content  = "${tls_self_signed_cert.root.cert_pem}"
```
Here and all other occurences, related to above, are not consistent with the other module: `--tls-ca-cert=${local.tls_dir}/ca.cert.pem`

natarajaya replied:
Fixed.

natarajaya commented:
@stepanstipl Thanks for the review! I addressed issues from your comments.

> My main question here is, how are the users supposed to handle migration

In short – yes, this change is disruptive, without additional action it will force TLS assets regeneration and make current Tiller installation unusable. TF state for `helm-tls` module needs to be rebuilt manually, I am currently working on migration procedure.

natarajaya commented:
@stepanstipl This is ready to be merged. I added migration plan to https://github.com/gpii-ops/gpii-infra/pull/280.

stepanstipl commented:
cool, thanks @natarajaya, Looks good 👍

mrtyler commented:
LGTM

Should this be proposed to upstream as well?

natarajaya commented:
> Should this be proposed to upstream as well?

@mrtyler I think so. Upstream master drifted pretty far away from our fork, I'll merge this, handle the migration and later figure out how to wrap this up for upstream PR.
